envoy/cilium_network_policy.h

```diff
@@ -0,0 +1,218 @@
+#include "envoy/config/subscription.h"
+#include "envoy/singleton/instance.h"
+#include "envoy/thread_local/thread_local.h"
+#include "envoy/http/header_map.h"
+
+#include "cilium/npds.pb.h"
+
+namespace Envoy {
+namespace Cilium {
+
+class NetworkPolicyMap : public Singleton::Instance,
+                         Config::SubscriptionCallbacks<cilium::NetworkPolicy>,
+                         public Logger::Loggable<Logger::Id::config> {
+public:
+  NetworkPolicyMap(const envoy::api::v2::core::ApiConfigSource& api_config_source,
+		   const LocalInfo::LocalInfo& local_info,
+		   Upstream::ClusterManager& cm, Event::Dispatcher& dispatcher,
+		   Stats::Scope &scope, ThreadLocal::SlotAllocator& tls);
+  NetworkPolicyMap(std::unique_ptr<Envoy::Config::Subscription<cilium::NetworkPolicy>>&& subscription,
+		   ThreadLocal::SlotAllocator& tls);
+  ~NetworkPolicyMap() {}
+
+  class PolicyInstance {
+  public:
+    PolicyInstance(uint64_t hash, const cilium::NetworkPolicy& proto)
+        : hash_(hash), policy_proto_(proto), ingress_(policy_proto_.ingress_per_port_policies()),
+          egress_(policy_proto_.egress_per_port_policies()) {}
+
+    uint64_t hash_;
+    const cilium::NetworkPolicy policy_proto_;
+
+  protected:
+    class HttpNetworkPolicyRule : public Logger::Loggable<Logger::Id::config> {
+    public:
+      HttpNetworkPolicyRule(const cilium::HttpNetworkPolicyRule& rule) {
+	ENVOY_LOG(trace, "Cilium L7 HttpNetworkPolicyRule():");
+	for (const auto& header: rule.headers()) {
+	  headers_.emplace_back(header);
+	  const auto& header_data = headers_.back();
+	  ENVOY_LOG(trace, "Cilium L7 HttpNetworkPolicyRule(): HeaderData {}={} ({})",
+		    header_data.name_.get(), header_data.value_, header_data.is_regex_ ? "regex" : "literal");
+	}
+      }
+
+      bool Matches(const Envoy::Http::HeaderMap& headers) const {
+	// Empty set matches any headers.
+	return Envoy::Router::ConfigUtility::matchHeaders(headers, headers_);
+      }
+
+      std::vector<Envoy::Router::ConfigUtility::HeaderData> headers_; // Allowed if empty.
+    };
+    
+    class PortNetworkPolicyRule : public Logger::Loggable<Logger::Id::config> {
+    public:
+      PortNetworkPolicyRule(const cilium::PortNetworkPolicyRule& rules) {
+	for (const auto& remote: rules.remote_policies()) {
+	  ENVOY_LOG(trace, "Cilium L7 PortNetworkPolicyRule(): Allowing remote {}", remote);
+	  allowed_remotes_.emplace(remote);
+	}
+	for (const auto& http_rule: rules.http_rules().http_rules()) {
+	  http_rules_.emplace_back(http_rule);
+	}
+      }
+
+      bool Matches(uint64_t remote_id, const Envoy::Http::HeaderMap& headers) const {
+	// Remote ID must match if we have any.
+	if (allowed_remotes_.size()) {
+	  bool matches = false;
+	  for (const auto& remote: allowed_remotes_) {
+	    if (remote == remote_id) {
+	      matches = true;
+	      break;
+	    }
+	  }
+	  if (!matches) {
+	    return false;
+	  }
+	}
+	// Empty set matches any payload
+	if (http_rules_.size() == 0) {
```

Comment:
I would restructure this check to be similar to the check above:
```
if (http_rules_.size() > 0) {
  bool matches = false;
  for (const auto& rule: http_rules_) {
    if (rule.Matches(headers)) {
      matches = true;
      break;
    }
  }
  if (!matches) {
    return false;
  }
}
return true;
```

Reply:
As there is no need for the variable here, I wrote this as:
```
	if (http_rules_.size() > 0) {
	  for (const auto& rule: http_rules_) {
	    if (rule.Matches(headers)) {
	      return true;
	    }
	  }
	  return false;
	}
	// Empty set matches any payload
	return true;
```